### PR TITLE
replace AsAny hack by trait upcasting :)

### PR DESCRIPTION
**Pull request by RalfJung**

### CODE REVIEW

**Comment by RalfJung** [comment]
Ah, this is actually wrong, I am upcasting the wrong thing. That's kind of subtle and easy to miss. (Cc @compiler-errors )

**Review by compiler-errors — APPROVED** [review]
r=me (if you're okay with non-miri approvals 😜)

**Comment by RalfJung** [comment]
> not &&dyn Tr

They were `Box<dyn Tr>`, which is very close to `&dyn Tr` (no double pointers!) so I was quite surprised to need `&**` here.

File: src/shims/unix/fs.rs [hunk 1]
```diff
@@ -72,6 +73,18 @@ pub trait FileDescriptor: std::fmt::Debug + helpers::AsAny {
...
+    #[inline(always)]
+    pub fn downcast_ref<T: Any>(&self) -> Option<&T> {
+        (self as &dyn Any).downcast_ref()
+    }
+
+    #[inline(always)]
+    pub fn downcast_mut<T: Any>(&mut self) -> Option<&mut T> {
+        (self as &mut dyn Any).downcast_mut()
+    }
+}
```
Comment (compiler-errors) [inline src/shims/unix/fs.rs:86]: Yeah this is the best approach!

**Pull request merged**